phoenix-core/src/main/java/org/apache/phoenix/compile/UpsertCompiler.java

```diff
@@ -787,15 +789,26 @@ public MutationState execute() throws SQLException {
                                 constantExpression.getDataType(), column.getDataType(), "expression: "
                                         + constantExpression.toString() + " in column " + column);
                         }
-                        if (!column.getDataType().isSizeCompatible(ptr, value, constantExpression.getDataType(),
+                        if ((column.getDataType() != PChar.INSTANCE) && // Char is truncated later
+                                !column.getDataType().isSizeCompatible(ptr, value, constantExpression.getDataType(),
```

Comment:
Is there a reason we wouldn't want to truncate the value for *all* fixed width types (i.e. column.getDataType().isFixedWidth)? In general, we try not to match on a particular type, but rather on categories of types. That way, when a new type gets added, it behaves a bit better.